chainstate/tx-verifier/src/transaction_verifier/utxos_undo_cache.rs

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::{btree_map::Entry, BTreeMap};
+
+use super::{error::ConnectTransactionError, TransactionSource};
+use chainstate_types::storage_result;
+use common::{
+    chain::{Block, Transaction},
+    primitives::Id,
+};
+use utxo::{UtxosBlockRewardUndo, UtxosBlockUndo, UtxosTxUndo};
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct UtxosBlockUndoEntry {
+    pub undo: UtxosBlockUndo,
+    // indicates whether this BlockUndo was fetched from the db or it's new
+    pub is_fresh: bool,
+}
```

Comment:
This seems to be a recurring pattern. Maybe it could be made generic with something like:
```rust
struct DbEntry<T> { data: T, is_fresh: bool }
```

And reuse that in various places. Not an issue for this PR though.